### PR TITLE
BUG: isin incorrectly casting ints to datetimes

### LINKED ISSUE #36621: BUG: isin - dtype conversions (Timestamp vs ints)
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

related: #34125 : could be closed in favor of that if it is too much of a duplicate.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
s = pd.Series(pd.date_range("jan-01-2013", "jan-05-2013"))
s.isin([1356998400000000000])
>>>
0     True
1    False
2    False
3    False
4    False
dtype: bool
```

#### Problem description

Problem arises here because `algos._ensure_dtype(s)` returns an `np.int64` array (that matches the data) and `dtype=<M8[ns]`. 

#### Expected Output

```py
0    False
1    False
2    False
3    False
4    False
dtype: bool
```

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #36621
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

De-duplication in follow-up.

### CODE REVIEW

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -430,6 +430,12 @@ def isin(comps: AnyArrayLike, values: AnyArrayLike) -> np.ndarray:
         # handle categoricals
         return cast("Categorical", comps).isin(values)
 
+    if needs_i8_conversion(comps):
+        # Dispatch to DatetimeLikeIndexMixin.isin
+        from pandas import Index
+
+        return Index(comps).isin(values)
```
Comment (jreback) [inline pandas/core/algorithms.py:437]: are you sure?

ensure_data and below does this

**Review by jreback — CHANGES_REQUESTED** [review]
this going to need perf tests as this is very sensitive

also i am surprised this is needed at all where you did it

File: pandas/core/algorithms.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/algorithms.py:437]: the way we call ensure_data ends up casting dt64 to int64

**Comment by jbrockmendel** [comment]
> this going to need perf tests as this is very sensitive

will run the appropriate asvs

**Comment by jbrockmendel** [comment]
actually, i think this can be simplified to just use _from_sequence following #37179

File: pandas/core/algorithms.py (hunk 1 above)
Comment (jreback) [reply pandas/core/algorithms.py:437]: why can't you fix it in _ensure_data then, I am not averse to dispatching to the EA itself, which is what we are already doing for Categorical, or is that the plan. If that is the plan, then I would think you could actually remove all off this and simply dispatch (e.g. most of this logic then would actually be handled by PandasArray or the EA as appropriate).

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/algorithms.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/algorithms.py:437]: There are a bunch of GH issues about isin (added a new label yesterday), many of which are about unwanted casting.  I haven't looked at them all closely tet.

> why can't you fix it in _ensure_data then

Even if we fix that (which yes, i'll take a look, but probably in a separate PR), we still need to dispatch to the `(Period|Timedelta|Datetime)Index` to get correct casting of the `other` (assuming we want `algos.isin(dtlike, other)` to match `dtlike.isin(other)`, which I for one do)

>  am not averse to dispatching to the EA itself, which is what we are already doing for Categorical, or is that the plan

ATM Categorical has `isin` but thats it.  #20617 suggests adding isin to EA, but thats definitely outside the scope of this PR.

**Comment by jbrockmendel** [comment]
gentle ping.  id like to see if we can share some code between this and `equals`

**Comment by jbrockmendel** [comment]
Found we don't have any asvs that hit this, so just added a few.  No change for cases with matching dtype, nice pickup for others:

```
       before           after         ratio
     [f823a859]       [9758b021]
     <ref-dtlike-indexers>       <bug-dti-isin>
-        691±30μs          143±3μs     0.21  series_methods.IsInDatetime64.time_isin_empty
-     1.23±0.08ms          141±8μs     0.11  series_methods.IsInDatetime64.time_isin_mismatched_dtype

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.

```

**Comment by jreback** [comment]
I guess as a short term measure this is ok as you fixed some bugs. the problem here is that you are bouncing between algos and the indexes, where everything should be in algos. this feels like a step backwards, if its tactical, then ok.

**Comment by jbrockmendel** [comment]
> the problem here is that you are bouncing between algos and the indexes, where everything should be in algos

its not unusual for core.algorithms functions to call EA methods.  There's no way the casting logic belongs in core.algorithms.

**Comment by jreback** [comment]
> > the problem here is that you are bouncing between algos and the indexes, where everything should be in algos
> 
> its not unusual for core.algorithms functions to call EA methods. There's no way the casting logic belongs in core.algorithms.

that's not what's happening here, these are calling Index methods. This seems quite unusual and a very odd depenency.
Why do you not simply move the logic in DTI.isin back to algorithms? what's the point of adding this depenedecy.?

**Comment by jbrockmendel** [comment]
> that's not what's happening here, these are calling Index methods

Good point, my mistake.  At some point this should be consolidated in the EA.  IIRC theres an issue about adding `isin` to the interface.

> Why do you not simply move the logic in DTI.isin back to algorithms? 

because the relevant logic is _casting_ logic that belongs in the DTA.  My current thought is that once we get the _from_sequence-strictness PR in we'll be able to de-duplicate some logic between isin+equals using that.

**Comment by jbrockmendel** [comment]
The new hashtables have me looking at algorithms._ensure_data, and I'm pretty confident that this will allow us to simplify it quite a bit

**Review by jreback — APPROVED** [review]
ok fair

**Pull request merged**